File: application/app/gateway/backend/BackendConnector.scala
package gateway.backend

import domain.model.BackendSettings
import domain.model.BackendType.BackendType
import gateway.events.InboundEvent

import scala.concurrent.Future

trait BackendConnector {
  type T <: BackendSettings

  def supports(backendType: BackendType): Boolean

  def sendEvent(event: InboundEvent, destination: String, settings: T): Future[Either[Exception, Unit]]
}

File: ws-gateway-application/src/main/java/com/cosmin/wsgateway/application/gateway/connector/BackendConnector.java
package com.cosmin.wsgateway.application.gateway.connector;

import com.cosmin.wsgateway.domain.Backend;
import com.cosmin.wsgateway.domain.BackendSettings;
import com.cosmin.wsgateway.domain.Event;
import reactor.core.publisher.Mono;

public interface BackendConnector<T extends BackendSettings> {
    boolean supports(Backend.Type type);

    Mono<Event> sendEvent(Event event, Backend<T> backend);
}
